### PR TITLE
WENGINES-4515 Setting up Interfaces for Github Checks (Part I)

### PR DESCRIPTION
**Pull request by Aayyush**
In order to support github checks, we need to make changes to the `CommitStatusUpdater`, which updates the status for a command/operation, and `PullUpdater`, which updates the pull request with the command/operation's output. In addition, we also need to keep the `statusId` in memory in all the command runners (where the status updates are performed): `Plan`, `Apply`, `PolicyCheck`, `ApprovePolicies` etc. Changes to the `StatusUpdater` and the `CommandRunner` needs to be provider-agnostic. So, I have tried to make minimally invasive changes to these structs while trying to keep them provider agnostic. 

On the client side, to keep things simple, we change the `UpdateStatus` method signature to return a `statusId` string as well. Github Client will create a new status check if the `statusId` is empty, and update it if it's not empty. All other vcs providers return an empty string for now, which keeps things same for other providers. 

Since the `pullUpdater` is provider-agnostic, we can't directly inject github checks specific logic to it. We could change the githubClient's `CreateComment()` internal implementation to update github checks with the output similar to how we do for `UpdateStatus` but, I don't think that's a good programming practice. Moreover, there's other structs that use this method to write comments to the pull request. So, I have added an `OuptutUpdaterProxy` which proxies the updateOutput request to the actual implementation based on the vcs provider.  For github, we have a `ChecksOutputUpdater` which calls the client's UpdateStatus method with the command result. 

### Note: 
In order to make keep this PR small, I will be adding the following changes in a follow up [**PR**](https://github.com/lyft/atlantis/pull/251/)
-  feature flags
- server flags
- delegate github status updates and inject into github client

### CODE REVIEW

File: server/events/vcs/github_client.go [hunk 1]
```diff
@@ -425,8 +425,18 @@ func (g *GithubClient) UpdateStatus(ctx context.Context, request types.UpdateSta
...
+
+	// TODO: For debug purposes, remove when implementing github checks
+	if request.StatusId == "" {
+		_, _, err := g.client.Repositories.CreateStatus(ctx, request.Repo.Owner, request.Repo.Name, request.Ref, status)
+		fmt.Printf("DEBUG: Creating status check: %v", status)
+		return "1", err
+	} else {
+		fmt.Printf("DEBUG: Updating status check for: %v", status)
+		_, _, err := g.client.Repositories.CreateStatus(ctx, request.Repo.Owner, request.Repo.Name, request.Ref, status)
+		return request.StatusId, err
```
Comment (samrabelachew) [inline server/events/vcs/github_client.go:437]: I'm a little confused here. What's the benefit of returning the same `request.StatusId` you passed in?

**Review by msarvar — COMMENTED** [review]
It is a bit unclear to me why we are passing around `statusId` in and our of interfaces? Can't you implement a new `GithubCheckClient` that implements `vcs.Client` interface and additionally can internally store statusId per workflow type(`plan`, `apply`, `policy_check`)? Seems like that will remove the need to change any interfaces and just keep the implementations inside the vcs client. Let me know if I'm missing something here.

**Comment by Aayyush** [comment]
> 



> It is a bit unclear to me why we are passing around `statusId` in and our of interfaces? Can't you implement a new `GithubCheckClient` that implements `vcs.Client` interface and additionally can internally store statusId per workflow type(`plan`, `apply`, `policy_check`)? Seems like that will remove the need to change any interfaces and just keep the implementations inside the vcs client. Let me know if I'm missing something here.

**Pull request closed without merging**